.github/workflows/ci.yml

```
@@ -3,38 +3,83 @@
+        if: always()
+        uses: actions/upload-artifact@v4
+        with:
+          name: jacoco
+          path: |
+            **/target/site/jacoco/jacoco.xml
+          if-no-files-found: ignore
 
   sonar:
+    needs: test
     runs-on: ubuntu-latest
+    if: (github.event_name == 'push' || github.event_name == 'pull_request') && needs.detect-changes.outputs.app == 'true'
```

actionlint (1.7.7)

property "detect-changes" is not defined in object type {test: {outputs: {}; result: string}}

(expression)